core/src/main/scala/kafka/server/AbstractFetcherManager.scala

```diff
@@ -191,6 +210,26 @@ abstract class AbstractFetcherManager[T <: AbstractFetcherThread](val name: Stri
   }
 }
 
+class FailedPartitions {
+  private val failedPartitions = new mutable.HashSet[TopicPartition]
+
+  def size: Int = synchronized {
+    failedPartitions.size
+  }
+
+  def add(partition: TopicPartition): Unit = synchronized {
+    failedPartitions += partition
+  }
+
+  def remove(partitions: Set[TopicPartition]): Unit = synchronized {
+    failedPartitions --= partitions
+  }
+
+  def getFailedPartitions: Set[TopicPartition] = {
```

Comment:
We're breaking the synchronization here. It seems like we just need this for a `contains` method. Maybe we could just expose `contains` and make it synchronized?